source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -475,11 +475,17 @@ std::string SslSocketInfo::ciphersuiteString() const {
   return SSL_CIPHER_get_name(cipher);
 }
 
-std::string SslSocketInfo::tlsVersion() const { return SSL_get_version(ssl_.get()); }
+const std::string& SslSocketInfo::tlsVersion() const {
+  if (!cached_tls_version_.empty()) {
```

Comment:
Now that we are passing this around via shared_ptr, I'm mildly concerned about potential thread safety with the cached strings. Should we be using read/write locking for the mutable variables? @lizan WDYT? (I can't think of a situation today in which this would break but I'm not sure about the future.)

Reply:
We don't pass this shared ptr across threads, I think this is fine for now. Do we plan to have multiple thread to process same connection?

Reply:
As a side note, SSL object from boringssl is not thread-safe.

Reply:
No there is no plan to pass connections. OK SGTM.